From d2f73e0e9ae168d4491a3427206ee41aa2e4616a Mon Sep 17 00:00:00 2001
From: Alexandru Costache <alexandru@balena.io>
Date: Wed, 23 Sep 2026 11:45:23 +0000
Subject: [PATCH] u-boot-compulab: adapt environment to balena bootloader

The changes attempt to boot the balena bootloader from the boot parition
falling back to the kernel in the root partition.

It also enables early console support.

Upstream-status: Inappropriate [Configuration]
Signed-off-by: Alex Gonzalez <alexg@balena.io>
---
 include/configs/compulab-imx8m-plus.h | 21 +++++++--------------
 1 file changed, 7 insertions(+), 14 deletions(-)

diff --git a/include/configs/compulab-imx8m-plus.h b/include/configs/compulab-imx8m-plus.h
index ae61b202aa6..ce196b69433 100644
--- a/include/configs/compulab-imx8m-plus.h
+++ b/include/configs/compulab-imx8m-plus.h
@@ -79,7 +79,7 @@
 	"mmcpart=" __stringify(CONFIG_SYS_MMC_IMG_LOAD_PART) "\0" \
 	"mmcroot=" CONFIG_MMCROOT " rootwait rw\0" \
 	"mmcautodetect=yes\0" \
-	"mmcargs=setenv bootargs console=${console} ${resin_kernel_root} ${os_cmdline} video=HDMI-A-1:D clk_ignore_unused "__stringify(BALENA_STAGE2)" \0 " \
+	"mmcargs=setenv bootargs console=${console} ${resin_kernel_root} ${root_opt} ${os_cmdline} video=HDMI-A-1:D clk_ignore_unused "__stringify(BALENA_STAGE2)" \0 " \
 	"loadbootscript=load mmc ${mmcdev}:${mmcpart} ${loadaddr} ${bsp_script};\0" \
 	"bootscript=echo Running bootscript from mmc ...; " \
 		"source\0" \
@@ -96,7 +96,7 @@
 				"echo WARN: Cannot load the DT; " \
 			"fi; " \
 		"fi;\0" \
-	"netargs=setenv bootargs console=${console} " \
+	"netargs=setenv bootargs " \
 		"root=/dev/nfs " \
 		"ip=dhcp nfsroot=${serverip}:${nfsroot},v3,tcp\0" \
 	"netboot=echo Booting from net ...; " \
@@ -123,15 +123,15 @@
 		"boot_part=1\0" \
 		"root_opt=rootwait ro fsck.mode=force fsck.repair=yes\0" \
 		"emmc_ul=setenv iface mmc; setenv dev ${resin_dev_index}; setenv part ${resin_root_part};" \
-		"setenv bootargs console=${console} ${resin_kernel_root} ${os_cmdline};\0" \
+		"run mmcargs; \0" \
 		"sd_ul=setenv iface mmc; setenv dev ${sd_dev}; setenv part ${boot_part};" \
-			"setenv bootargs console=${console} root=${sd_root} ${root_opt};\0" \
+			"setenv bootargs console=${console} root=${sd_root} ${root_opt}  "__stringify(BALENA_STAGE2)"; \0" \
 		"usb_ul=usb start; setenv iface usb; setenv dev ${usb_dev}; setenv part ${boot_part};" \
-			"setenv bootargs console=${console} root=${usb_root} ${root_opt};\0" \
+			"setenv bootargs root=${usb_root} ${root_opt} "__stringify(BALENA_STAGE2)"; \0" \
 		"ulbootscript=load ${iface} ${dev}:${part} ${scriptaddr} ${bsp_script};\0" \
 		"ulrunbootscript=echo Running bootscript from ${iface} ${dev}:${part} ...; " \
 				"source ${scriptaddr}\0" \
-		"ulimage=if load ${iface} ${resin_dev_index}:${resin_root_part} ${zip_addr} boot/${image}; then unzip ${zip_addr} ${loadaddr}; run balena_kernel_load_crc_save; else false; fi; \0" \
+		"ulimage=if load ${iface} ${dev}:${resin_boot_part} ${zip_addr} ${image}; then unzip ${zip_addr} ${loadaddr}; run balena_kernel_load_crc_save; else if load ${iface} ${dev}:${resin_root_part} ${zip_addr} boot/${image}; then unzip ${zip_addr} ${loadaddr}; run balena_kernel_load_crc_save; else false; fi; fi; \0" \
 		"ulfdto=setenv fdto1file; for fdto1file in ${fdtofile}; do "\
 			    "echo loading ${fdto1file}; "\
 			    "load ${iface} ${dev}:${part} ${fdto_addr_r} ${fdto1file} "\
@@ -139,14 +139,7 @@
 			    "&& fdt resize 0x8000 "\
 			    "&& fdt apply ${fdto_addr_r};"\
 		"done; true;\0"\
-		"ulfdt=if test ${boot_fdt} = yes || test ${boot_fdt} = try; then " \
-			"load ${iface} ${resin_dev_index}:${resin_root_part} ${fdt_addr_r} boot/${fdtfile} && run balena_fdt_load_crc_save;" \
-			"if env exists fdtofile;then "\
-					"run ulfdto; "\
-				"else "\
-					"true; "\
-				"fi; "\
-		"fi;\0" \
+		"ulfdt=if load ${iface} ${dev}:${resin_boot_part} ${fdt_addr_r} ${fdtfile}; then run balena_fdt_load_crc_save; else if load ${iface} ${dev}:${resin_root_part} boot/${fdt_addr_r} ${fdtfile}; then run balena_fdt_load_crc_save; else false; fi; fi; \0" \
 		"bootlist=emmc_ul\0" \
 		"bsp_bootcmd=echo Running BSP bootcmd ...; " \
 		"for src in ${bootlist}; do " \
-- 
2.34.1

